### PR TITLE
added more sort files and further testing for those classes

### PR DESCRIPTION
**Pull request by jhcoll**

### CODE REVIEW

File: src/main/java/com/scottlogic/AuthorPostSorter.java [hunk 1]
```diff
@@ -1,16 +1,22 @@
...
     @Override
-    public List<UserPost> sort(List<UserPost> inputList) {
-        if(inputList != null){
-                inputList.sort(Comparator.comparing(UserPost::getAuthor));
+    public List<UserPost> sort(List<UserPost> inputList, SortOrder SortOrder) {
+        if (inputList != null) {
+            List<UserPost> copyInputList = new ArrayList<>(inputList);
+            copyInputList.sort(Comparator.comparing(UserPost::getAuthor));
+            if (SortOrder == com.scottlogic.SortOrder.DESC) {
+                copyInputList.sort(Collections.reverseOrder(Comparator.comparing(UserPost::getAuthor)));
```
Comment (jluxford-scottlogic) [inline src/main/java/com/scottlogic/AuthorPostSorter.java:16]: At the moment this is sorting the list twice in the case of `DESC`

File: src/main/java/com/scottlogic/AuthorPostSorter.java [hunk 2]
```diff
@@ -1,16 +1,22 @@
...
+            }
+            return copyInputList;
         }
-        return inputList;
+        return null;
```
Comment (jluxford-scottlogic) [inline src/main/java/com/scottlogic/AuthorPostSorter.java:20]: In general having a code block inside an `if` statment where the outside just returns is better done by having the inverse condition. So using `if (inputList == null) {return null}` makes the code look a lot cleaner. It means that the condition and action are close together and the rest of the code doesn't end up half way across the screen

File: src/main/java/com/scottlogic/AuthorPostSorter.java [hunk 3]
```diff
@@ -1,16 +1,22 @@
...
 
 public class AuthorPostSorter implements PostSorter {
 
     @Override
-    public List<UserPost> sort(List<UserPost> inputList) {
-        if(inputList != null){
-                inputList.sort(Comparator.comparing(UserPost::getAuthor));
+    public List<UserPost> sort(List<UserPost> inputList, SortOrder SortOrder) {
+        if (inputList != null) {
+            List<UserPost> copyInputList = new ArrayList<>(inputList);
```
Comment (jluxford-scottlogic) [inline src/main/java/com/scottlogic/AuthorPostSorter.java:13]: `copyInputList` is a good name to begin with but after a few lines the naming breaks down a little as it implies that this will always be a copy of the input list which isn't true. Maybe something like `mutableInputList`, `sortableList`, `outputList` or something that is true throughout the programme might be a better shout.

File: src/main/java/com/scottlogic/DatePostSorter.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+public class DatePostSorter implements PostSorter {
+
+    @Override
+    public List<UserPost> sort(List<UserPost> inputList, SortOrder SortOrder) {
```
Comment (jluxford-scottlogic) [inline src/main/java/com/scottlogic/DatePostSorter.java:11]: Good work, in general similar comments to above apply to this class as well

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 5]
```diff
@@ -10,66 +8,109 @@
...
+        List<UserPost> inputList = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> expected = Arrays.asList(userPost1, userPost2, userPost3);
+
+        List<UserPost> actual = new AuthorPostSorter().sort(inputList, SortOrder.DESC);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void sort_unsorted2Num_sorted2Num() {
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:97]: In general you want least complicated to most complicated tests as you go down the page, sorting two items is less complicated than sorting three so should probably be higher up

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 6]
```diff
@@ -10,66 +8,109 @@
...
-import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
 
 class AuthorPostSorterTest {
+    UserPost userPost1 = new UserPost("Joe Bloggs",
+            OffsetDateTime.of(2020, 1, 3, 7, 12, 3, 0, ZoneOffset.UTC),
+            "Hello World!", 2);
+
+    UserPost userPost2 = new UserPost("Joe Bloggs",
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:19]: (Minor) Nice catch to have the same value to check for stability. My only concern with this is that stability is an edge case and so should probably be given a separate couple of tests that explicitly state that they are testing the stability. This is because tests often want to be as clear as possible from the naming and the code as to the reason that they failed, if your standard sorting 3 desc tests failed but it was a stability issue then it might take a lot longer to work out what the issue was as you might think it was desc in general rather than stability

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java [hunk 7]
```diff
@@ -10,66 +8,109 @@
...
 
-        List<UserPost> actual = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> inputList = Arrays.asList(userPost1, userPost2, userPost3);
         List<UserPost> expected = Arrays.asList(userPost3, userPost1, userPost2);
 
-        actual = new AuthorPostSorter().sort(actual);
+        List<UserPost> actual = new AuthorPostSorter().sort(inputList, SortOrder.ASC);
 
         assertEquals(expected, actual);
     }
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/AuthorPostSorterTest.java:116]: Another edge case that might be worth testing is the mutability of the input, by doing an assert on a copy of the input after the sort has run you can check that it is the same as before. I would advise doing this in a different test and not checking that the expected output is the same as the actual output within the same test to further the goal of making the problem really clear from the test that is failing. For example if this started failing because of a stability issue on the output you might also think a mutation issue is happening

File: src/test/java/com/scottlogic/DatePostSorterTest.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        List<UserPost> inputList = Arrays.asList(userPost1, userPost3, userPost2);
+        List<UserPost> expected = Arrays.asList(userPost1, userPost2, userPost3);
+
+        List<UserPost> actual = new DatePostSorter().sort(inputList, SortOrder.ASC);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void sort_unsorted3Lower_sorted3Lower() {
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/DatePostSorterTest.java:65]: This test probably isn't needed as whether the name is lowercase shouldn't have any impact on how the date is sorting things. You could claim it as an edge case but it looks more like a copy paste oversight to me so thought I'd bring it up.

File: src/test/java/com/scottlogic/DatePostSorterTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        List<UserPost> inputList = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> expected = Arrays.asList(userPost3, userPost2, userPost1);
+
+        List<UserPost> actual = new DatePostSorter().sort(inputList, SortOrder.DESC);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void sort_withSameDate_sortedSameDate() {
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/DatePostSorterTest.java:97]: This is a really good example of splitting out the stability aspect for the asc and desc tests nice work

File: src/test/java/com/scottlogic/DatePostSorterTest.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+                OffsetDateTime.of(2020, 3, 12, 13, 22, 12, 0, ZoneOffset.UTC),
+                "An example of a post \nwith lines breaks.", 3);
+
+        List<UserPost> inputList = Arrays.asList(userPost1, userPost2, userPost3);
+        List<UserPost> expected = Arrays.asList(userPost3, userPost1, userPost2);
+
+        List<UserPost> actual = new DatePostSorter().sort(inputList, SortOrder.DESC);
+
+        assertEquals(expected, actual);
+    }
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/DatePostSorterTest.java:138]: On sneaky edge cases mutability could be useful and do you know how it behaves with time zones?

File: src/test/java/com/scottlogic/LikesPostSorterTest.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class LikesPostSorterTest {
+    UserPost userPost1 = new UserPost("Joe Bloggs",
```
Comment (jluxford-scottlogic) [inline src/test/java/com/scottlogic/LikesPostSorterTest.java:14]: A general things about naming here which is a bit more obvious with this case is it is often nice to have the names reflect information about the object. At the moment `userPostX` doesn't mean `X` likes, this means that when looking through the tests you have to keep coming back up here to make sure the orderings are correct. If instead `userPostX` had `X` likes then the output list would be really obvious it was in the correct order and you could see the impact of the rearranged input list as well.

**Review by jluxford-scottlogic — APPROVED** [review]
Nice work the tests have covered most of the edge cases and look like a really solid suite. A few comments on general style, testing conventions and a few extra cases that might be worth testing for. All looking good so far, I'm a little more confident with java and a bit of a stickler with tests so I might pick up on a bit more for the next leg of the code if that is good with you?

File: src/test/java/com/scottlogic/AuthorPostSorterTest.java (hunk 7 above)
Comment (jhcoll) [reply src/test/java/com/scottlogic/AuthorPostSorterTest.java:116]: Sorry could you try and clarify this? Are you saying that I should write a test that parses a sorted list, and checks that the output has not changed from he input?
